File: src/SimpleLevelEditor.Formats.Types/Rgba.fs
namespace SimpleLevelEditor.Formats.Types

open System.Numerics

[<Struct>]
type Rgba =
    struct
        val mutable R: byte
        val mutable G: byte
        val mutable B: byte
        val mutable A: byte

        new(r, g, b, a) = { R = r; G = g; B = b; A = a }

        member this.ToVector4() = Vector4(float32 this.R / 255.0f, float32 this.G / 255.0f, float32 this.B / 255.0f, float32 this.A / 255.0f)

        static member Default = Rgba(0uy, 0uy, 0uy, 0uy)
    end

File: src/SimpleLevelEditor.Formats.Core/Rgba.cs
using System.Numerics;

namespace SimpleLevelEditor.Formats.Core;

public record struct Rgba
{
	public byte R;
	public byte G;
	public byte B;
	public byte A;

	public Rgba(byte r, byte g, byte b, byte a)
	{
		R = r;
		G = g;
		B = b;
		A = a;
	}

	public Vector4 ToVector4()
	{
		return new Vector4(R / 255f, G / 255f, B / 255f, A / 255f);
	}

	public string ToDisplayString()
	{
		return $"{R.ToDisplayString()} {G.ToDisplayString()} {B.ToDisplayString()} {A.ToDisplayString()}";
	}
}
